src/Controller/ViewerController.php

```diff
@@ -55,22 +53,6 @@ public function editor(Media $media, $edit = false) {
             'closebutton' => 'true',
         ];
 
-        $user = \Drupal::currentUser();
-        $permissions = CoolUtils::getUserPermissions($user);
-
-        if (!$permissions['is_viewer']) {
-            $error_msg = 'Authentication failed.';
-            \Drupal::logger('cool')->error($error_msg);
-            return new Response(
-                $error_msg,
-                Response::HTTP_FORBIDDEN,
-                ['content-type' => 'text/plain']
-            );
-        }
-
-        /* Make sure that the user is a collaborator if edit is true */
-        $edit = $edit && $permissions['is_collaborator'];
-
```

Comment:
How do we check about the write permission? The code above make sure that if you get an "editor" but don't have write permission as per drupal, then you get bounced back to a read-only. It also reject if you don't have viewing permission. This is better than WOPI errors in Collabora online.

Reply:
In the new version the access is checked at the route level.
This means that, if you attempt to access `'/cool/edit/{media}'` path when you only have "preview" permission, you get an access denied page, instead of the "readonly" fallback.

Also, now it is technically possible to have an "edit" permission without having a "preview" permission.
Such a user would be able to access the `'/cool/edit/{media}'` path, but not the `'/cool/view/{media}'` path.
It would be up to the admin / webmaster / site builder to only assign "edit" to roles that also have "preview".

How important is the scenario with the fallback?
One use case I could imagine is that user A with edit access passes a link to user B who has only read/preview access. In that case we would want user B to see the preview page.

If this is really important, we would need to move part of the the access logic back into the controller.
On the other hand, having the permission on the route level allows Drupal to hide the "edit" link for people with no access, whenever it would be displayed in a menu or in "tabs". (Currently no such places exist for these specific links, where the visibility would depend on the route access check.)

Reply:
> Permissions are set at the entity level

> In the new version the access is checked at the route level.

This sounds contradictory, but in fact both of these statements are true :)
The access check happens at the route level, so before the controller method is called.
However, the route access check then invokes the entity access system, with `_entity_access: 'media.edit in collabora'`.

Reply:
Having the access check on route level, and not in the controller, is a standard practice in Drupal for most routes, providing a clean separation between controller and access check. With all the consequences mentioned above, e.g. when you visit `/node/123/edit`, if you don't have edit access, you get an access denied, not a fallback to the node view page.

Exceptions are technically possible, e.g. for API urls that should be accessed by a script or service, and where a default "access denied" response would not be suitable.

In this case we would prefer to stick to the default Drupal behavior, at least for now.
This should be ok because we also hide the link for user without edit access.